### ISSUE TITLE
Create a new responsive modal or any way to convert an already open modal on another type of modal

### ISSUE THREAD

**Issue by JGeek00** [flutter/flutter#133946]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

I am using modal bottom sheets and full screen modals on my application. The problem that I currently have is that I feel that on big screens, the modal bottom sheet and the fullscreen modal are a bit useless because they are designed to be used on mobile screens and they waste too much space on tablet and desktop sizes. So I want to do something similar to what iOS does. On the iPhone the modal fills almost the entire screen, but on the iPad the modal is a big sheet on the center of the screen. I want to have something similar to that but with the material styling.

Now, if you create a fullscreen modal, a normal modal, or a bottom sheet, you don't have any way to convert it into another type of modal when the user resizes the window, or when the user rotates the device.

![image](https://github.com/flutter/flutter/assets/47545344/2e0a68a6-2d27-49d7-944d-401282e80f47)

### Proposal

It would be great to have on the Material package something similar to the fullscreen modal on iOS. A responsive modal that on mobile is a fullscreen modal and on tablet or desktop is a normal modal on the center of the screen. The problem that we currently have is that the modal and the fullscreen modal are triggered on different ways, and when the modal is open and the user resizes the window or rotates the device, you cannot convert one modal into the other one.

Normal modal:
````
showModal(
  context: context, 
  builder: (ctx) => Modal()
)
````

Fullscreen modal:
````
Navigator.push(context, MaterialPageRoute(
  fullscreenDialog: true,
  builder: (BuildContext context) => Container()
))
````

Something similar to this but with the material interface.
![image](https://github.com/flutter/flutter/assets/47545344/2e0a68a6-2d27-49d7-944d-401282e80f47)

Also it would be great to have something similar to what I described previously but with the modal bottom sheet and the normal modal. A modal that on mobile is a modal bottom sheet, and on tablet and desktop is a normal modal on the center of the screen.

My suggestion:
````
showResponsiveFullScreenModal(
  context: context,
  breakingWidth: 800,
  builder: (context) => Container()
)
````
````
showResponsiveModalBottomSheet(
  context: context,
  breakingWidth: 800,
  builder: (context) => Container()
)
````

**Comment by dam-ease** [comment]
Hi @JGeek00. Thanks for proposing this. Can you check out this package https://pub.dev/packages/wolt_modal_sheet to see if it fits into what you are trying to achieve??

**Comment by JGeek00** [comment]
> Hi @JGeek00. Thanks for proposing this. Can you check out this package https://pub.dev/packages/wolt_modal_sheet to see if it fits into what you are trying to achieve??

Yes, my objective is to do something similar to what that library does, but the problem is that it gives you a strict structure to build the modal instead of an empty container where you can put the structure that you want. It would be great to have that behavior built in the default material package but on the most simple way:
````
showResponsiveModal(
  context: context,
  configuration: ResponsiveModalConfiguration(
    breakingWidth: 800,
    small: (Widget child) => Container(     // the child property contains what the user added to the builder property
      decoration: BoxDecoration(
        define properties to make a container look like a bottom sheet
      ),
      child: child,
    ),
    big: (Widget child) => Dialog(     // the child property contains what the user added to the builder property
      child: child,
    )
  ),
  builder: (context) => Container()
)

// Also it would be necessary a new FullScreenDialog() widget with the looking of the full screen dialog
````
and what you add to the builder is wrapped inside the bottom sheet or the dialog depending on the screen width.

I forked that library and tried to adapt it to my needs but it's too complex for me. I don't have that much experience working with Flutter. It would be great to have that built into the material package to make it available to other devs that want to use responsive modals.


@iapicca as I said before, I don't know how to do that.

**Comment by HansMuller** [comment]
This sounds like a good idea. We would welcome a PR!

**Comment by CodeDoctorDE** [comment]
Hi, I have a `ResponsiveDialog` built for my own use case: https://github.com/LinwoodDev/dart_pkgs/blob/main/packages/material_leap/lib/src/dialogs/responsive.dart.
It solves my usecase and should be pretty universal.
I'm also open to contribute this to the new `material_ui` package. Should I first build some kind of design doc?
